### PR TITLE
Implementation_of_user_api

### PR DESCRIPTION
**Pull request by AnkushM123**
I have implemented user api with validation and swagger implementation

### CODE REVIEW

File: apis/login.js [hunk 1]
```diff
@@ -0,0 +1,239 @@
...
+*                 type: boolean
+*                isDeleted:
+*                 type: boolean
+*       400:
+*         description: Bad request
+*/
+
+const registerUser = async (req, res) => {
+  const validatePassword = (password) => {
+    const regex = /^(?=.*[a-z])(?=.*[A-Z])(?=.*\d)(?=.*[@$!%*?&])[A-Za-z\d@$!%*?&]{8,}$/;
```
Comment (RavirajSuryavanshi) [inline apis/login.js]: Please explain the regex logic here in comments.

File: apis/login.js [hunk 2]
```diff
@@ -0,0 +1,239 @@
...
+    return (regex.test(password));
+  }
+
+  if (!req.body.name.trim() || !req.body.address.trim() || !req.body.mobile.trim() || !req.body.gender.trim() || req.body.roles.length == 0 || !req.body.email.trim() || !req.body.password.trim() || req.body.age < 1) {
```
Comment (RavirajSuryavanshi) [inline apis/login.js]: What will happen if the body.name/address/mobile etc filled is not passed in the request?

File: apis/login.js [hunk 3]
```diff
@@ -0,0 +1,239 @@
...
+    return res.status(400).send("name,address,age,mobile,gender,role,email and password are required fields");
```
Comment (RavirajSuryavanshi) [inline apis/login.js]: The validation/error message should be specific and not generic. Please ensure we provide information which is relevant to the user and not confusing

File: apis/login.js [hunk 4]
```diff
@@ -0,0 +1,239 @@
...
+  }
+
+  if (req.body.age < 0 || req.body.age > 60) {
```
Comment (RavirajSuryavanshi) [inline apis/login.js]: In the above if control we are checking age < 1 and here we are checking for age < 0. What is the reason behind diff checks here?

File: apis/login.js [hunk 5]
```diff
@@ -0,0 +1,239 @@
...
+  if (!validator.isEmail(req.body.email)) {
+    return res.status(400).send("please enter valid email");
+  }
+ 
+  if (!validatePassword(req.body.password)) {
+    return res.status(400).send("Password must contain atleast one lower,one upper,one special character,one digit,no blank spaces and length must be between 8-20 characters");
+  }
+
+  const hashedPassword = await bcrypt.hash(req.body.password, 10);
+  const add = JSON.parse(req.body.address)
```
Comment (RavirajSuryavanshi) [inline apis/login.js]: The name of the variables should be meaningful. "Add" can have multiple meanings

File: apis/login.js [hunk 6]
```diff
@@ -0,0 +1,239 @@
...
+    name: req.body.name,
+    address: {
+      flat_details: add.flat_details,
+      landmark: add.landmark,
+      area: add.area 
+    },
+    age: req.body.age,
+    mobile: req.body.mobile,
+    gender: req.body.gender,
+    roles: req.body.roles.split(','),
```
Comment (RavirajSuryavanshi) [inline apis/login.js]: We are accepting the body.roles as an array then why are we spliting it?

File: apis/login.js [hunk 7]
```diff
@@ -0,0 +1,239 @@
...
+  })
+
+  let data = await userService.createUser(employee);
+  res.send(data);
+}
+
+const authenticateToken = (req, res, next) => {
+  const token = req.headers.authorization?.split(' ')[1];
+  if (!token) {
+    return res.status(401).json({ error: 'Unauthorized: Token missing' });
```
Comment (RavirajSuryavanshi) [inline apis/login.js]: Any sort of response should come from resourcing files and should not be a hard coded string

File: apis/login.js [hunk 8]
```diff
@@ -0,0 +1,239 @@
...
+*/
+
+const login = async (req, res) => {
+  try {
+    if (!validator.isEmail(req.body.email)) {
+      return res.status(400).send("please enter valid email and password");    
+    }
+
+    const data = await userService.getUserByEmail(req.body.email);
+    if (data.length==0) {
```
Comment (RavirajSuryavanshi) [inline apis/login.js]: What if the data is null here?

File: apis/login.js [hunk 9]
```diff
@@ -0,0 +1,239 @@
...
+      return res.status(400).send({ message: 'Invalid username or password' });
+    }
+
+    const isMatch = await bcrypt.compare(req.body.password, data[0].password);
+    const user = { "id": data[0]._id }
+
+    if (isMatch) {
+      const token = jwt.sign(user, process.env.secretKey, { expiresIn: '1h' });
```
Comment (RavirajSuryavanshi) [inline apis/login.js]: expiresIn should not be hard coded

File: apis/user.js [hunk 10]
```diff
@@ -1,15 +1,353 @@
...
+*         description: Invalid Id
+*         content:
+*           text/plain:
+*             schema:
+*               type: string
+*               example: 'Invalid Id format'
+*/
+
+const editUser = async (req, res) => {
+  let isValidObjectId = (id) => {
```
Comment (RavirajSuryavanshi) [inline apis/user.js]: What is the validation that we are trying apply here?

File: apis/user.js [hunk 11]
```diff
@@ -1,15 +1,353 @@
...
+
+    const data = await userService.editUser(req.params.id, employee);
+
+    if (data.modifiedCount === 1) {    
+      return res.send("user updated successfully")
     }
-    else {
-      res.status(404).json("No data found");
+    else {  
+      return res.status(404).json("No data found");
```
Comment (RavirajSuryavanshi) [inline apis/user.js]: Why are we returning 404?

File: apis/user.js [hunk 12]
```diff
@@ -1,15 +1,353 @@
...
+*           text/plain:
+*             schema:
+*               type: string
+*               example: 'please enter valid email'
+*/
+
+const getUserByEmail = async (req, res) => {
+
+  if (validator.isEmail(req.body.email)) {
+    const _email = await userService.getUserByEmail(req.body.email);
```
Comment (RavirajSuryavanshi) [inline apis/user.js]: why we are using const here?

File: apis/user.js [hunk 13]
```diff
@@ -1,15 +1,353 @@
...
+*         description: Invalid Email
+*         content:
+*           text/plain:
+*             schema:
+*               type: string
+*               example: 'please enter valid email'
+*/
+
+const changePassword = async (req, res) => {
+  const validatePassword = (password) => {
```
Comment (RavirajSuryavanshi) [inline apis/user.js]: this code is repeated many time

**Review by RavirajSuryavanshi — CHANGES_REQUESTED** [review]

File: core/services/user-service.js [hunk 14]
```diff
@@ -8,4 +12,12 @@ const createUser = async function (user) {
...
 
-module.exports = { getUserByEmail, createUser };
\ No newline at end of file
+const editUser = async function (id, employee) {
+    return userModel.updateOne({ _id: id }, { $set: employee })
+}
+
+const changePassword = async function (email, password) {
+    return userModel.findOneAndUpdate({ email: email }, { $set: { password: password } })
+}
```
Comment (ix-pankajk) [inline core/services/user-service.js]: apply filter on id

**Review by ix-pankajk — CHANGES_REQUESTED** [review]

**Review by ix-pankajk — APPROVED** [review]

**Pull request merged**